arangod/Aql/OptimizerRules.cpp

```diff
@@ -4436,6 +4437,125 @@ void arangodb::aql::inlineSubqueriesRule(Optimizer* opt,
   opt->addPlan(std::move(plan), rule, modified);
 }
 
+
+
+ExecutionNode* getOnlyParent(ExecutionNode* node){
+  if(!node){ return nullptr; }
+  auto parents = node->getParents();
+  if(parents.size() != 1) {
+    LOG_DEVEL << "more than one parent";
+    return nullptr;
+  }
+  auto rv = parents[0];
+  TRI_ASSERT(rv != node);
+  if(rv == node) {
+    LOG_DEVEL << "node is it's own parent";
+    return nullptr;
+  }
+  return rv;
+}
+
+ExecutionNode* findInParents(ExecutionNode * const startNode
+                            ,ExecutionNode::NodeType typeToFind
+                            ,std::vector<ExecutionNode::NodeType> const& allowedTypes
+                            ,bool readOnlySubQueries = true
+                            ){
+
+  ExecutionNode* currentNode = getOnlyParent(startNode); //move down from start node
+
+  while(currentNode){
+    auto currentType = currentNode->getType();
+
+    if(currentType == typeToFind){
+      return currentNode; // current node matches the search
+    }
+
+    auto found = std::find(allowedTypes.begin(), allowedTypes.end(), currentType);
+    if(found != allowedTypes.end()){
+      if (readOnlySubQueries &&
+          currentType == ExecutionNode::SUBQUERY &&
+          static_cast<SubqueryNode*>(currentNode)->isModificationQuery()
+         ) { return nullptr; }
+      currentNode = getOnlyParent(currentNode); //advance
+    } else {
+      LOG_DEVEL_IF( currentType != 18 ) << currentType << " is not allowed!";
+      return nullptr; // type not allowed
+    }
+  }
+  // no valid current node - terminate unsuccessful search
+  return nullptr;
+}
+
+
+void arangodb::aql::optimizeClusterLimitsToShardsRule(Optimizer* opt,
+                                             std::unique_ptr<ExecutionPlan> plan,
+                                             OptimizerRule const* rule) {
+
+  TRI_ASSERT(arangodb::ServerState::instance()->isCoordinator());
+  bool wasModified = false;
+
+  SmallVector<ExecutionNode*>::allocator_type::arena_type s;
+  SmallVector<ExecutionNode*> nodes{s};
+  std::vector<ExecutionNode::NodeType> const types = {ExecutionNode::GATHER};
+  plan->findNodesOfType(nodes, types, true);
```

Comment:
could simply use the non-vector method here (first overload in the following snippet), as the search vector only contains a single node type:
```
156   /// @brief find nodes of a certain type
157   void findNodesOfType(SmallVector<ExecutionNode*>& result,
158                        ExecutionNode::NodeType,
159                        bool enterSubqueries);
160 
161   /// @brief find nodes of a certain types
162   void findNodesOfType(SmallVector<ExecutionNode*>& result,
163       std::vector<ExecutionNode::NodeType> const&, bool enterSubqueries); 
```